### ISSUE TITLE
checksum function assumes little endian

### ISSUE THREAD

**Issue by bendk** [mozilla/uniffi-rs#1393]
[Building Firefox on big-endian systems is failing after we added our first UniFFI component to it](https://bugzilla.mozilla.org/show_bug.cgi?id=1797714).  It seems like the cultprit is our [checksum function](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_meta/src/lib.rs#L128-L132).  It should probably take the lowest 16 bits rather than the last 16 bits.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/UNIFFI-209)
┆Issue Number: UNIFFI-209
┆Link To Issue: https://mozilla-hub.atlassian.net/browse/UNIFFI-209

**Comment by cuviper** [comment]
If you mean this line:

```rust
    (hasher.finish() & 0x000000000000FFFF) as u16
```

That much is independent of endianness, taking the least significant bits regardless of byte order.
(The mask is also redundant, because `as u16` will truncate that anyway.)

However, `DefaultHasher` is not endian-safe -- each of its `Hasher::write_u16`/etc. methods use the trait-default behavior of `self.write(&i.to_ne_bytes())`. If you pick a different hasher type, then you should also make sure that the `T: Hash` values you feed in are also consistent. We face this challenge in `rustc` itself too...

**Comment by cuviper** [comment]
While you're at it, I would also beware of 32/64-bit differences, especially `usize`.

**Comment by bendk** [comment]
@cuviper thanks for the help here.  Do you think `rustc-hash` could work for us?

We have a couple types where we implement `Hash` ourselves, but I think those should still be consistent since they just call their field's hash() methods.

**Comment by cuviper** [comment]
No, I think `rustc-hash` would have the same problems with endianness and pointer size. A better match for your needs is the compiler's `StableHasher`, but I guess you'll have to copy that because it's not published as a crate.

https://github.com/rust-lang/rust/blob/4c736a21ae317086871094b1671d6535be593502/compiler/rustc_data_structures/src/stable_hasher.rs#L12-L21

**Comment by cuviper** [comment]
FWIW, I brought this up on Rust Zulip, and I think that conversation might be helpful to you as well:
https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/StableHasher.20.28lazy.20pre-ACP.29

**Comment by mhammond** [comment]
I wonder if this is actually a problem for uniffi in general, or a problem with how we are using it in mozilla-central? While I can't find an open bug for this, I recall that we ultimately want that binding generation to happen at build time rather than having those files checked in - and I think @bendk started trying to make this work but ended up with problems having uniffi-bindgen available at build time. If we had this, this problem wouldn't have surfaced.

For our other existing uniffi consumers, this also isn't a problem in practice - there is an implied expectation that bindings are generated at build time, and this checksum mechanism is really just to try and help avoid this expectation falling down - eg, where the component changes and is built, but the foreign bindings build didn't happen. I don't think there's an expectation that you can generate the foreign bindings once and have them used with various different build targets.

tl;dr - while this capability might be nice in an abstract sense, is the actual problem here with uniffi, or just with the fact the generated bindings are checked in to m-c?

**Comment by cuviper** [comment]
It's definitely easier if everything using the hash is generated at the same time!

**Comment by bendk** [comment]
@mhammond I think that's correct.  But if I'm thinking of a way to get the build running again, it might be easier to make the hash stable than to make the generation happen at build time.  But then again, I definitely agree that having the generation happen at build time would be good for other reasons too.

**Comment by bendk** [comment]
Fixed by #1427.
